### PR TITLE
Add the possibility to inject a custom RecoveryState factory to IndexStorePlugin implementations

### PR DESCRIPTION
**Pull request by fcofdez**
Add a custom factory for recovery state into `IndexStorePlugin` that
allows different implementors to provide its own `RecoveryState`
implementation.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 1]
```diff
@@ -742,6 +745,30 @@ public synchronized void verifyIndexMetadata(IndexMetadata metadata, IndexMetada
...
+            recoveryTargetService,
+            recoveryListener,
+            repositoriesService,
+            onShardFailure,
+            globalCheckpointSyncer,
+            retentionLeaseSyncer);
+    }
+
     @Override
     public IndexShard createShard(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java:773]: Can we lose this override? It looks only to be used in one test that could use the default recovery state factory anyway.

**Review by DaveCTurner — COMMENTED** [review]
Thanks @fcofdez, I left one question.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 2]
```diff
@@ -742,6 +745,30 @@ public synchronized void verifyIndexMetadata(IndexMetadata metadata, IndexMetada
...
+            final ShardRouting shardRouting,
+            final PeerRecoveryTargetService recoveryTargetService,
+            final PeerRecoveryTargetService.RecoveryListener recoveryListener,
+            final RepositoriesService repositoriesService,
+            final Consumer<IndexShard.ShardFailure> onShardFailure,
+            final Consumer<ShardId> globalCheckpointSyncer,
+            final RetentionLeaseSyncer retentionLeaseSyncer,
+            final DiscoveryNode targetNode,
+            final DiscoveryNode sourceNode) throws IOException {
+        IndexService indexService = indexService(shardRouting.index());
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java:759]: Can we assert that `indexService` exists?

File: server/src/main/java/org/elasticsearch/plugins/IndexStorePlugin.java [hunk 3]
```diff
@@ -55,4 +60,26 @@
...
+
+    /**
+     * The {@link RecoveryStateFactory} mappings for this plugin. When an index is created the recovery type setting
+     * {@link org.elasticsearch.index.IndexModule#INDEX_RECOVERY_TYPE_SETTING} on the index will be examined and either use the default
+     * or looked up among all the recovery state factories from {@link IndexStorePlugin} plugins.
+     *
+     * @return a map from recovery type to an recovery state factory
+     */
+    default Map<String, RecoveryStateFactory> getRecoveryStateFactories() {
+        return Collections.emptyMap();
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/plugins/IndexStorePlugin.java:83]: Can we add a test to ensure that a plugin cannot override the default recovery state factory? And that duplicates are not allowed too?

**Review by tlrx — COMMENTED** [review]
Thanks for the PR spliting Francisco. I've left some comments

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 1 above)
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/indices/IndicesService.java:773]: Do you mean removing this method from `AllocatedIndices` interface?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/IndicesService.java:773]: Yes

**Comment by fcofdez** [comment]
@elasticmachine update branch

File: server/src/test/java/org/elasticsearch/plugins/IndexStorePluginTests.java [hunk 4]
```diff
@@ -92,4 +127,20 @@ public void testDuplicateIndexStoreFactories() {
...
+                                                                                    BarCustomRecoveryStore.class)));
+        if (JavaVersion.current().compareTo(JavaVersion.parse("9")) >= 0) {
+            assertThat(e, hasToString(matches(
+                "java.lang.IllegalStateException: Duplicate key recovery-type \\(attempted merging values " +
+                    "org.elasticsearch.plugins.IndexStorePluginTests\\$RecoveryFactory@[\\w\\d]+ " +
+                    "and org.elasticsearch.plugins.IndexStorePluginTests\\$RecoveryFactory@[\\w\\d]+\\)")));
+        } else {
+            assertThat(e, hasToString(matches(
+                "java.lang.IllegalStateException: Duplicate key " +
+                    "org.elasticsearch.plugins.IndexStorePluginTests$RecoveryFactory@[\\w\\d]+")));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/plugins/IndexStorePluginTests.java:143]: I think we shouldn't be asserting that our recovery factories have a `toString()` that returns `org.elasticsearch.plugins.IndexStorePluginTests$RecoveryFactory@[\\w\\d]+`: the fact that the message contains the duplicated key (and the string `Duplicate key`) should be enough IMO.

Bit of a pain the message doesn't contain the key in Java 8 but such is life.

**Review by DaveCTurner — APPROVED** [review]
LGTM; one small (optional) comment about the strength of the assertions in the new test.

**Review by tlrx — APPROVED** [review]
LGTM - Thanks for the additional tests

**Pull request merged**